crypto/rand/drbg_lib.c

```diff
@@ -282,21 +285,27 @@ int RAND_DRBG_instantiate(RAND_DRBG *drbg,
     }
 
     drbg->state = DRBG_ERROR;
+
+    if (drbg->min_noncelen > 0 && drbg->get_nonce == NULL) {
```

Comment:
The condition `drbg->min_noncelen > 0` holds precisely in the case when a derivation function is used, which is the case you are dealing with here. But that is not obvious from the if-condition. To make it more clear, I would replace line 289 by: 

```
    if ((drbg->flags & RAND_DRBG_FLAG_CTR_NO_DF) == 0) {
        if (drbg->get_nonce == NULL) {
            ...
        }
    }
```

Reply:
This if-block also needs an explaining comment which cites the section from the NIST document that suggests  this way of obtaining the nonce..

Reply:
I'm also not sure whether this is the optimal way of solving the nonce-problem. Entropy might be a scarce resource on the device. Wouldn't the date+counter variant which I mentioned before be 'cheaper'?

Reply:
The nonce is only used for the initialization, so it's not using that much more entropy. It's also not really making sense to me why it needs 50% more entropy and how you can replace that with a timestamp and a counter that is never going to increase, so I prefer to be on the safe side.

I also don't like to make assumptions that if we're using a DF that we require a nonce, and if we're not we don't. If we require a nonce, the minimum nonce length is > 0, which is what the code now checks.